src/test/scala/freeeventsourcing/ProcessTestSupport.scala

```diff
@@ -0,0 +1,140 @@
+  def runFromWithResult(expectations: Expectation*)(toResult: Any) =
+    ProcessMatcher(expectations.toList, r ⇒
+      if (r == toResult) Xor.right(()) else Xor.left(s"produced non matching result: ${r} instead of ${toResult}"))
+
+  /** Helper methods to construct the expectations. */
+  object Expect {
+    import Expectation._
+    def awaitEvent[S <: WithEventType](selector: S)(result: S#Event) =
+      ExpectAwaitEvent(selector, result)
+    def waitUntil(instant: Instant) = ExpectWaitUntil(instant)
+    def commandSuccessful[A <: Aggregate, C <: A#Command](aggregateType: A, aggregate: A#Id, command: C) =
+      ExpectCommand(aggregateType, aggregate, command, Xor.right(()))
+    def commandFailed[A <: Aggregate, C <: A#Command, E](aggregateType: A, aggregate: A#Id, command: C)(result: E)(
+      implicit
+      i: Inject[C#Error, E]
+    ) =
+      ExpectCommand(aggregateType, aggregate, command, Xor.left(i(result)))
+    def end = ExpectEnd
+  }
+
+  case class ProcessMatcher(expectations: List[Expectation], checkResult: Any ⇒ String Xor Unit) extends Matcher[M[_]] {
+    def apply(process: M[_]) = {
+      val x = process.foldMap(Implementation.Transform)
+      val y = x
+        .run(expectations)
+        .leftMap(error ⇒ s"did not match the expectations: ${error}")
+        .flatMap {
+          case (Nil, r)       ⇒ Xor.right(r)
+          case (remaining, _) ⇒ Xor.left(s"remaining expectations: ${remaining.mkString(", ")}")
+        }
+        .flatMap(checkResult)
+      new MatchResult(
+        y.isRight,
+        y.fold(identity, _ ⇒ "?"),
+        s"matched the process expectations but should not have"
+      )
+    }
+  }
+
+  type Action[+A] = ProcessAction[BC, A]
+  type M[A] = ProcessMonad[BC, A]
+
+  sealed trait Expectation
+  object Expectation {
+    case class ExpectAwaitEvent[S <: WithEventType](selector: S, result: S#Event) extends Expectation
+    case class ExpectWaitUntil(instant: Instant) extends Expectation
+    case class ExpectCommand[A <: Aggregate, Cmd <: A#Command](
+      aggregateType: A, aggregate: A#Id, command: A#Command, result: Cmd#Error Xor Unit
+    ) extends Expectation
+    case object ExpectEnd extends Expectation
+  }
+
+  private[this] object Implementation {
+    import Expectation._
+
+    type OrFail[A] = Xor[String, A]
+    type Expectations[A] = StateT[OrFail, List[Expectation], A]
+    object Expectations {
+      def next(received: String): Expectations[Expectation] = lift {
+        case h :: t ⇒ Xor.right((t, h))
+        case Nil    ⇒ Xor.left(s"No more expectations, but got ${received}")
+      }
+
+      def check[A](action: Action[A]): Expectations[A] = next(action.toString).flatMap { exp ⇒
+        compare.lift((action, exp)).getOrElse {
+          lifted(Xor.left(s"Expected ${exp} but got ${action}"))
+        }
+      }
+
+      private[this] def compare[A]: PartialFunction[(Action[A], Expectation), Expectations[A]] = {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Cyclomatic complexity of 15 exceeds max of 10](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2580073129/issues/source?bid=3298199&fileBranchId=3298202#l95)